statsmodels/stats/contingency_tables.py

```diff
@@ -0,0 +1,1215 @@
+        e_stat = um * vn / n_obs
+        v_stat = (u2m - um**2 / n_obs) * (v2n - vn**2 / n_obs) / (n_obs - 1)
+        sd_stat = np.sqrt(v_stat)
+
+        zscore = (stat - e_stat) / sd_stat
+        pvalue = 2 * stats.norm.cdf(-np.abs(zscore))
+
+        b = _bunch()
+        b.stat = stat
+        b.stat_e0 = e_stat
+        b.stat_sd0 = sd_stat
+        b.zscore = zscore
+        b.pvalue = pvalue
+        return b
+
+
+    @cache_readonly
+    def marginal_probabilities(self):
+        """
+        Return the estimated row and column marginal distributions.
+        """
+        n = self.table.sum()
+        row = self.table.sum(1) / n
+        col = self.table.sum(0) / n
+
+        if isinstance(self.table_orig, pd.DataFrame):
+            row = pd.Series(row, self.table_orig.index)
+            col = pd.Series(col, self.table_orig.columns)
+
+        return row, col
+
+
+    @cache_readonly
+    def independence_probabilities(self):
+        """
+        Estimated cell probabilities under independence.
+        """
+        row, col = self.marginal_probabilities
+        itab = np.outer(row, col)
+
+        if isinstance(self.table_orig, pd.DataFrame):
+            itab = pd.DataFrame(itab, self.table_orig.index,
+                                self.table_orig.columns)
+
+        return itab
+
+
+    @cache_readonly
+    def fittedvalues(self):
+        """
+        Fitted values under independence.
+        """
+        probs = self.independence_probabilities
+        fit = self.table.sum() * probs
+        return fit
+
+
+    @cache_readonly
+    def pearson_resids(self):
+        """
+        The Pearson residuals.
+        """
+        fit = self.fittedvalues
+        resids = (self.table - fit) / np.sqrt(fit)
+        return resids
+
+
+    @cache_readonly
+    def standardized_resids(self):
+        """
+        Residuals with unit variance.
+        """
+        row, col = self.marginal_probabilities
+        sresids = self.resids / np.sqrt(np.outer(1 - row, 1 - col))
+        return sresids
+
+
+    @cache_readonly
+    def chi2_contribs(self):
+        """
+        The contribution of each cell to the chi^2 statistic.
+        """
+        return self.pearson_resids**2
+
+
+    @cache_readonly
+    def local_log_oddsratios(self):
+        """
+        Returns the local log odds ratios.
+
+        The local log odds ratios are calculated for each 2x2 subtable
+        formed from adjacent rows and columns.
+        """
+
+        ta = self.table.copy()
+        a = ta[0:-1, 0:-1]
+        b = ta[0:-1, 1:]
+        c = ta[1:, 0:-1]
+        d = ta[1:, 1:]
+        tab = np.log(a) + np.log(d) - np.log(b) - np.log(c)
+        rslt = np.empty(self.table.shape, np.float64)
+        rslt *= np.nan
+        rslt[0:-1, 0:-1] = tab
+
+        if isinstance(self.table_orig, pd.DataFrame):
+            rslt = pd.DataFrame(rslt, index=self.table_orig.index,
+                                columns=self.table_orig.columns)
```

Comment:
A possible problem of caching the DataFrame is that internal calculations that use this will have either DataFrame or numpy arrays to work with. That makes the code more fragile if there are calculations where the behavior of DataFrames and numpy arrays differ.